### PR TITLE
clarify CPU and memory limit enforcement differences

### PR DESCRIPTION
**Pull request by haircommander**
Clairify the container runtime and kubelet do not enforece memory limits, the kernel does. Also clarify that workloads won't always be OOM killed, and more clearly spell out the differences between memory and CPU limits

### CODE REVIEW

**Comment by sftim** [comment]
@kubernetes/sig-node-pr-reviews FYI

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 1]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
+and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
+by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits
+are enforced differently.
 
-Limits can be implemented either reactively (the system intervenes once it sees a violation)
-or by enforcement (the system prevents the container from ever exceeding the limit). Different
-runtimes can have different ways to implement the same restrictions.
+`CPU` limits are enforced by CPU throttling. When a container approaches
+its `CPU` limit, the kernel will restrict access to the CPU corresponding to the
+container's limit.
```
Comment (sftim) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:45]: `cpu` (lower case) please @haircommander; the original is wrong, but we should fix it. In the actual API, it's `cpu`.

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 2]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
+
+`memory` limits are enforced by the kernel with OOM (out of memory) kills. When
+a pod uses more than its `memory` limit, the kernel may terminate it. However,
+if there is available memory on the node, the kernel may choose not to terminate
+the offending pod.
```
Comment (sftim) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:50]: 1. What happens to memory allocation attempts -  `sbrk()` and friends - that would assign memory above the limit, if they were to succeed? Ideally, let's be clear about where the enforcement does and doesn't happen.

1. The container runtime or kernel might apply a higher limit than you request. For example, if you specify a limit of 2097159 bytes (a shade over 2MiB), the kernel is likely to round things up to a whole page size. That's if your container runtime even lets you request a limit that small!

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 3]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
+are enforced differently.
 
-Limits can be implemented either reactively (the system intervenes once it sees a violation)
-or by enforcement (the system prevents the container from ever exceeding the limit). Different
-runtimes can have different ways to implement the same restrictions.
+`CPU` limits are enforced by CPU throttling. When a container approaches
+its `CPU` limit, the kernel will restrict access to the CPU corresponding to the
+container's limit.
+
+`memory` limits are enforced by the kernel with OOM (out of memory) kills. When
```
Comment (saschagrunert) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:47]:
```suggestion
`memory` limits are enforced by the kernel with out of memory (OOM) kills. When
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 4]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
+Both `CPU` and `memory` limits are applied by the kubelet (and
+{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
+and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
+by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits
+are enforced differently.
 
-Limits can be implemented either reactively (the system intervenes once it sees a violation)
-or by enforcement (the system prevents the container from ever exceeding the limit). Different
-runtimes can have different ways to implement the same restrictions.
+`CPU` limits are enforced by CPU throttling. When a container approaches
```
Comment (saschagrunert) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:43]: Can we link a follow-up resource on how CPU throttling works in linux?

**Review by saschagrunert — COMMENTED** [review]
Great update, thank you!

**Comment by dipesh-rawat** [comment]
@haircommander Whenever you get a chance, please review the feedback from the reviewers and respond to them accordingly. Thanks!

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 4 above)
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:43]:
```suggestion
`cpu` limits are enforced by [CPU throttling](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling) When a container approaches
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 5]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
+its `CPU` limit, the kernel will restrict access to the CPU corresponding to the
```
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:44]:
```suggestion
its `cpu` limit, the kernel will restrict access to the CPU corresponding to the
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 4 above)
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:43]:
```suggestion
`cpu` limits are enforced by [CPU throttling](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling) When a container approaches
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 6]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
-If you set a `memory` limit of 4GiB for that container, the kubelet (and
-{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}) enforce the limit.
-The runtime prevents the container from using more than the configured resource limit. For example:
-when a process in the container tries to consume more than the allowed amount of memory,
-the system kernel terminates the process that attempted the allocation, with an out of memory
-(OOM) error.
+Both `CPU` and `memory` limits are applied by the kubelet (and
+{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
+and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
+by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits
```
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:40]:
```suggestion
Both `cpu` and `memory` limits are applied by the kubelet (and
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `cpu` and `memory` limits
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 2 above)
Comment (tengqm) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:50]: I don't agree to this change related to memory limits. It is not accurate and it will lead to more confusion, because memory limit enforcement is a pretty complex topic.

Comment (network-charles) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:50]: Is it because the kubelet or container runtime uses a [cgroup driver](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#cgroup-drivers) to interact with the OS’s underlying cgroup system to enforce a limit, hence, any limit created by a different cgroup does not matter?

Comment (tengqm) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:50]: @network-charles Did you read this? https://kubernetes.io/blog/2023/05/05/qos-memory-resources/

Comment (network-charles) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:50]: I just did. It's similar to my last comment but with an added info on the “memory.high” feature supported by cgroupv2. It's clear now.

We may need to clarify that the kernel uses the “memory. high” feature supported on cgroupv2 to throttle the container. Also, the statement that it applies to a pod is wrong, according to the blog, it only applies at the container cgroups level.

Comment (tengqm) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:50]: So ... the point is ... a Pod may and may not get killed when it is using more memory than its limit.
Can we rephrase this to be more accurate?

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 7]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
+are enforced differently.
```
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:41]: > However, `CPU` and `memory` limits
 are enforced differently.

The above statement is wrong. You already answered correctly in lines 37 - 38, yet contradicted your answer in lines 40 - 41. There's no need to throw the kernel into the conversation here, it ends up confusing the reader.

Only two things are missing from the existing documentation.
1. An outline that the kubelet and container runtime [communicates](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#cgroup-drivers) with **control groups** to enforce the limit.
2. A configured `memory.high` setting in cgroupv2 can prevent a container from getting killed by an OOM error when it exceeds its limit.

Since these are what you are trying to add, I suggest you use the existing documentation and only include what is missing instead of changing all the sentences.

```suggestion
If you set a `memory` limit of 4GiB for that container, the kubelet (and
 {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}) communicates with control groups to enforce the limit.
 The runtime prevents the container from using more than the configured resource limit. For example:
 when a process in the container tries to consume more than the allowed amount of memory,
 the system kernel terminates the process that attempted the allocation, with an out of memory
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 4 above)
Comment (network-charles) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:43]: I removed the follow-up resource. I don't think changing this line is necessary. See this [review](https://github.com/kubernetes/website/pull/46222/files#r1793695009).

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 8]
```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
...
```
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:42]: I'll suggest you leave this as it is.
```suggestion
(OOM) error.
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 1 above)
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:45]: This too.
I'll suggest you reference a resource showing how kubernetes implements CPU throttling, a concept similar to the `memory.high` settings in [cgroupv2](https://kubernetes.io/blog/2023/05/05/qos-memory-resources/#how-it-works).
```suggestion
Limits can be implemented either reactively (the system intervenes once it sees a violation)
or by enforcement (the system prevents the container from ever exceeding the limit). Different
 runtimes can have different ways to implement the same restrictions.
```

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 2 above)
Comment (network-charles) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:50]:
```suggestion
In [cgroup v2](https://kubernetes.io/docs/concepts/architecture/cgroups/), if you configure the `memory.high` [settings](https://docs.kernel.org/admin-guide/cgroup-v2.html#memory) and a
pod uses more than its memory limit, the kernel will throttle the memory of the pod.
However, this throttling will only occur if the node has enough available memory.
```

**Comment by divya-mohan0209** [comment]
@haircommander : Gentle reminder!

**Comment by haircommander** [comment]
I finally dug out some time to respond! I have updated and clarified some sections

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 1 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: I tried updating so whenever the limit is specified I used lowercase

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 2 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: I added a note about memoryQOS feature, it's not something we should document heavily because the effort is stalled in SIG node due to issues

Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: > What happens to memory allocation attempts - sbrk() and friends - that would assign memory above the limit, if they were to succeed? Ideally, let's be clear about where the enforcement does and doesn't happen.

it's unfortunately not so easy. When the kernel enforces OOM kills, it usually is in response to a memory allocation. so sometimes, sbrk and friends will trigger an oom kill of the process, and other times it will succeed.

> The container runtime or kernel might apply a higher limit than you request. For example, if you specify a limit of 2097159 bytes (a shade over 2MiB), the kernel is likely to round things up to a whole page size. That's if your container runtime even lets you request a limit that small!

I don't know if we need to get to this level of detail here, but I'm open to suggestions

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 3 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: good point!

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 4 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: > Can we link a follow-up resource on how CPU throttling works in linux?

I can't find a good one that isn't on some companies website. the cgroup documentation doesn't really go into depth on how it works. Any suggestions?

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 7 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: I have clarified this section, I think it's important to mention who is actually doing the enforcing and how. I did mention MemoryQoS but I don't think we should go in depth because we're not moving forward with that effort now.

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 8 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:54]: this line was dropped because the paragraph was reworked, so I think we should keep it dropped

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 2 above)
Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: we stalled on this effort so i don't think we should mention it deeply

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 7 above)
Comment (network-charles) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: Nice!

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 8 above)
Comment (network-charles) [reply content/en/docs/concepts/configuration/manage-resources-containers.md:54]: Alright

File: content/en/docs/concepts/configuration/manage-resources-containers.md (hunk 4 above)
Comment (network-charles) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: I previously used this one from [Wikipedia](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling) since its open source.

Comment (haircommander) [reply content/en/docs/concepts/configuration/manage-resources-containers.md]: hm maybe I'm using throttling wrong in this case. That article looks like it's describing behavior of the hardware of scaling its density up and down to match workloads. In this case, this is really CPU time slicing, where a process gets slices of the CPU time according to its weight, and is limited in going above that CPU time limit

**Comment by network-charles** [comment]
The clarification is lovely, @haircommander, I have a better understanding now. I'm looking forward to feedback from others.

File: content/en/docs/concepts/configuration/manage-resources-containers.md [hunk 9]
```diff
@@ -28,22 +28,37 @@ that system resource specifically for that container to use.
...
 
-Limits can be implemented either reactively (the system intervenes once it sees a violation)
-or by enforcement (the system prevents the container from ever exceeding the limit). Different
-runtimes can have different ways to implement the same restrictions.
+{{< note >}}
+There is an alpha feature `MemoryQoS` which attempts to add more preemptive
+limit enforcement for memory (as opposed to reactive enforcement by the OOM
+killer). However, this effort is
+[stalled](https://github.com/kubernetes/enhancements/tree/a47155b340/keps/sig-node/2570-memory-qos#latest-update-stalled)
+due to a potential livelock situation a memory hungry can cause.
```
Comment (sftim) [inline content/en/docs/concepts/configuration/manage-resources-containers.md:60]: Did you mean: a memory hungry **application**?

**Review by sftim — COMMENTED** [review]
/approve

This change makes the docs much better.
